internal/rpc/core/env.go

```diff
@@ -239,6 +242,41 @@ func (env *Environment) StartService(ctx context.Context, conf *config.Config) (
 		cfg.WriteTimeout = conf.RPC.TimeoutBroadcastTxCommit + 1*time.Second
 	}
 
+	// If the event log is enabled, subscribe to all events published to the
+	// event bus, and forward them to the event log.
+	if lg := env.EventLog; lg != nil {
+		// TODO(creachadair): This is kind of a hack, ideally we'd share the
+		// observer with the indexer, but it's tricky to plumb them together.
+		// For now, use a "normal" subscription with a big buffer allowance.
+		// The event log should always be able to keep up.
+		const subscriberID = "event-log-subscriber"
+		sub, err := env.EventBus.SubscribeWithArgs(ctx, tmpubsub.SubscribeArgs{
+			ClientID: subscriberID,
+			Query:    query.All,
+			Limit:    1 << 16, // essentially "no limit"
+		})
+		if err != nil {
+			return nil, fmt.Errorf("event log subscribe: %w", err)
+		}
+		go func() {
+			// N.B. Use background for unsubscribe, ctx is already terminated.
```

Comment:
It sort feels like the eventbus and the subscription objects should manage their own lifescycles, rather than having to handle it at this point.

I'd love, at some point, to have functions that start services have a convention of either: letting those objects manage their own lifecylces or returning a `func(context.Context) error` so that we could avoid skipping shutdown ops or being relatively confident that shutdown wouldn't deadlock. That day is not today.

Reply:
I generally agree. The tricky bit is that the eventbus has no notion of subscriptions of its own—it just forwards subscription requests to the pubsub. The pubsub is a separate service, however, and both it and its subscriptions have separate lifecycles.

If we were planning to hang onto the existing subscription mechanism, it would probably be worthwhile to do something better with its lifecycle. Closing down the pubsub _does_ terminate all running subscriptions right now, but otherwise they do not obey the context they're given when constructed.

I tried to fix that in what I thought was the obvious way: "Just make it track the context passed to Subscribe". However, a bunch of existing code assumes it can keep the sub after the context it passed in expires. Fixing that was a much bigger lift, and I didn't think it was worthwhile at the time.

Reply:
(This is further complicated by the fact that the Unsubscribe mechanism is the only way to terminate a subscription explicitly. The Subscription channel never closes—and fixing that exploded the world when I tried it—which means the client basically has to do the call manually, as here)

Reply:
yeah, I was messing with this locally just to see if I was totally off my rocker, and while I think this model is good, I think the code you have is more than fine.

Reply:
> yeah, I was messing with this locally just to see if I was totally off my rocker, and while I think this model is good, I think the code you have is more than fine.

You're right, that's how it _should_ work. It mainly turned out that getting there from how it _does_ work was harder than I anticipated, so I kinda cut a corner 😄